pkg/alertmanager/distributor_test.go

```diff
@@ -0,0 +1,297 @@
+			// reach this point the 3rd AM may not have received the request yet.
+			// To avoid flaky test we retry until we hit the desired state within a reasonable timeout.
+			test.Poll(t, time.Second, true, func() interface{} {
+				totalReqCount := 0
+				for _, a := range ams {
+					reqCount := a.requestsCount(route)
+					// AM should not get duplicate requests.
+					require.True(t, reqCount <= 1, "duplicate requests %d", reqCount)
+					totalReqCount += reqCount
+				}
+
+				return c.expectedTotalCalls == totalReqCount
+			})
+		})
+	}
+
+}
+
+func prepare(t *testing.T, numAM, numHappyAM, replicationFactor int) (*Distributor, []*mockAlertmanager, func()) {
+	ams := []*mockAlertmanager{}
+	for i := 0; i < numHappyAM; i++ {
+		ams = append(ams, newMockAlertmanager(t, i, true))
+	}
+	for i := numHappyAM; i < numAM; i++ {
+		ams = append(ams, newMockAlertmanager(t, i, false))
+	}
+
+	// Use a real ring with a mock KV store to test ring RF logic.
+	amDescs := map[string]ring.InstanceDesc{}
+	amByAddr := map[string]*mockAlertmanager{}
+	for i, a := range ams {
+		amDescs[a.myAddr] = ring.InstanceDesc{
+			Addr:                a.myAddr,
+			Zone:                "",
+			State:               ring.ACTIVE,
+			Timestamp:           time.Now().Unix(),
+			RegisteredTimestamp: time.Now().Add(-2 * time.Hour).Unix(),
+			Tokens:              []uint32{uint32((math.MaxUint32 / numAM) * i)},
+		}
+		amByAddr[a.myAddr] = ams[i]
+	}
+
+	kvStore := consul.NewInMemoryClient(ring.GetCodec())
+	err := kvStore.CAS(context.Background(), RingKey,
+		func(_ interface{}) (interface{}, bool, error) {
+			return &ring.Desc{
+				Ingesters: amDescs,
+			}, true, nil
+		},
+	)
+	require.NoError(t, err)
+
+	amRing, err := ring.New(ring.Config{
+		KVStore: kv.Config{
+			Mock: kvStore,
+		},
+		HeartbeatTimeout:  60 * time.Minute,
+		ReplicationFactor: replicationFactor,
+	}, RingNameForServer, RingKey, nil)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), amRing))
+	test.Poll(t, time.Second, numAM, func() interface{} {
+		return amRing.InstancesCount()
+	})
+
+	cfg := &MultitenantAlertmanagerConfig{}
+	flagext.DefaultValues(cfg)
+
+	d, err := NewDistributor(cfg, amRing, newMockAlertmanagerClientFactory(amByAddr), util_log.Logger, prometheus.NewRegistry())
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), d))
+
+	return d, ams, func() {
+		require.NoError(t, services.StopAndAwaitTerminated(context.Background(), d))
+	}
+}
+
+type mockAlertmanager struct {
+	alertmanagerpb.AlertmanagerClient
+	grpc_health_v1.HealthClient
+	t *testing.T
+	// receivedRequests is map of route -> statusCode -> number of requests.
+	receivedRequests map[string]map[int]int
+	mtx              sync.Mutex
+	myAddr           string
+	happy            bool
+}
+
+func newMockAlertmanager(t *testing.T, idx int, happy bool) *mockAlertmanager {
```

Comment:
Looks a bad smell passing `testing.T` to a mock. I would remove it. See related comments below.